File: apps/ios/Voiid/Voiid/Models/Story.swift
//
//  Story.swift
//  Voiid
//
//  Stories domain models: the E2EE envelopes that cross the wire (inside the Double
//  Ratchet, so the server never sees them) and the local UI models rendered from
//  StoryStore. A story is the sendMedia flow with the recipient set widened from one
//  peer to a chosen audience — see docs/STORIES_PROTOCOL.md.
//
//  MediaRef is DELIBERATELY reused verbatim from ChatEngine.swift (do not redefine it):
//  the media key rides inside the envelope exactly as it does for a media message, so
//  it never leaves E2E.
//

import Foundation

// MARK: - Wire envelopes (authenticated plaintext, encrypted per target device)

/// The story key envelope (§1.3 step 5). This whole struct is the plaintext of a
/// Double-Ratchet message, so GCM authenticates every field. There is no AAD on any
/// AEAD in e2e-core, so story_id / author_id / expiry are bound HERE and the receiver
/// MUST validate them against the feed row before showing anything (§1.5).
struct StoryEnvelope: Codable {
    var v: Int = 1
    var t: String = "story"
    let story_id: String
    let author_id: String
    let created_at: Int64        // epoch ms
    let expires_at: Int64        // author's claim; server value wins for freshness
    let media: MediaRef          // the EXISTING MediaRef shape, verbatim
    var caption: String = ""
    var durationMs: Int?         // video only; drives the segment timer
    var width: Int?
    var height: Int?
    var allowsReplies: Bool = true
}

// StoryReplyEnvelope is defined in Networking/ChatEngine.swift, NOT here: ChatEngine.swift
// is compiled into the VoiidNSE extension target (which excludes this app-only model layer),
// and ChatEngine.decodeEnvelope decodes a story reply while the NSE decrypts an inbound push.
// Every wire type ChatEngine touches lives beside it for that reason.

/// A view receipt's ratchet plaintext (§4.3). The viewer identity lives INSIDE this
/// encrypted payload — the server never gets a viewer_user_id column.
struct StoryViewEnvelope: Codable {
    var v: Int = 1
    var t: String = "story_view"
    let story_id: String
    let viewer_id: String
    let viewed_at: Int64         // epoch ms
}

// MARK: - Local UI models

/// Download lifecycle for a story's plaintext blob. Mirrors the DB `download_state`.
enum StoryDownloadState: String {
    case none, downloading, ready, failed, gone
}

/// One story as the UI renders it, decoded from a `stories` row.
struct Story: Identifiable, Equatable {
    let id: String
    let authorId: String
    let authorDeviceId: String?
    let isMine: Bool
    let createdAt: Date
    let expiresAt: Date
    let media: MediaRef
    let caption: String
    let durationMs: Int?
    let width: Int?
    let height: Int?
    let allowsReplies: Bool
    var viewedAt: Date?
    var localPath: String?
    var downloadState: StoryDownloadState

    var isExpired: Bool { expiresAt <= Date() }
    var isViewed: Bool { viewedAt != nil }

    /// Segment duration for the viewer's timer (§8.4). Static images get 5s; video
    /// plays for its natural length capped at 30s.
    var segmentDuration: TimeInterval {
        if let ms = durationMs, ms > 0 { return min(Double(ms) / 1000.0, 30.0) }
        return 5.0
    }

    static func == (l: Story, r: Story) -> Bool { l.id == r.id && l.viewedAt == r.viewedAt && l.downloadState == r.downloadState && l.localPath == r.localPath }
}

/// An author's set of unexpired stories — one row/cell in the tray, never one per story.
struct StoryContext: Identifiable, Equatable {
    let authorId: String
    var stories: [Story]        // ascending by createdAt (viewer steps through in order)

    var id: String { authorId }
    var isMine: Bool { stories.first?.isMine ?? false }
    var newest: Story? { stories.max(by: { $0.createdAt < $1.createdAt }) }
    /// A context is "unviewed" if any of its stories is unseen — drives the accent ring.
    var hasUnviewed: Bool { stories.contains { !$0.isViewed } }
    /// Index of the first unseen story so the viewer opens where the user left off.
    var firstUnviewedIndex: Int { stories.firstIndex(where: { !$0.isViewed }) ?? 0 }

    static func == (l: StoryContext, r: StoryContext) -> Bool {
        l.authorId == r.authorId && l.stories == r.stories
    }
}

File: apps/android/app/src/main/java/com/voiid/app/model/Story.kt
package com.voiid.app.model

import com.voiid.app.net.ChatEngine
import kotlinx.serialization.Serializable

/**
 * Stories domain models + the on-the-wire envelopes.
 *
 * TIME UNITS: everything here is epoch MILLIS (like the rest of the app models). The Room columns
 * hold SECONDS; the only place that converts is [com.voiid.app.store.StoryLocalStore].
 *
 * THE PRIVACY MODEL, in one line so nobody re-litigates it in a PR: the media blob is encrypted
 * ONCE on-device with a fresh random AES-256-GCM key and PUT to R2; the key travels inside a
 * ~400-byte [StoryEnvelope] encrypted ONCE PER RECIPIENT DEVICE over that device's existing
 * Double Ratchet session. The server sees ciphertext, opaque object keys, and the set of device
 * ids that received key material (= the audience — metadata we do NOT claim to hide). It never
 * sees the media key, the real mime, the caption, the dimensions, or the bytes.
 */

/** Download lifecycle of the ciphertext blob for one story on THIS device. */
enum class StoryDownloadState { NONE, DOWNLOADING, READY, FAILED, GONE }

fun String.toStoryDownloadState(): StoryDownloadState = when (this) {
    "downloading" -> StoryDownloadState.DOWNLOADING
    "ready" -> StoryDownloadState.READY
    "failed" -> StoryDownloadState.FAILED
    "gone" -> StoryDownloadState.GONE
    else -> StoryDownloadState.NONE
}

fun StoryDownloadState.wire(): String = when (this) {
    StoryDownloadState.DOWNLOADING -> "downloading"
    StoryDownloadState.READY -> "ready"
    StoryDownloadState.FAILED -> "failed"
    StoryDownloadState.GONE -> "gone"
    StoryDownloadState.NONE -> "none"
}

/** Upload lifecycle of one of OUR outgoing stories (optimistic post). */
enum class StoryUploadState { NONE, UPLOADING, SENT, FAILED }

fun String.toStoryUploadState(): StoryUploadState = when (this) {
    "uploading" -> StoryUploadState.UPLOADING
    "sent" -> StoryUploadState.SENT
    "failed" -> StoryUploadState.FAILED
    else -> StoryUploadState.NONE
}

fun StoryUploadState.wire(): String = when (this) {
    StoryUploadState.UPLOADING -> "uploading"
    StoryUploadState.SENT -> "sent"
    StoryUploadState.FAILED -> "failed"
    StoryUploadState.NONE -> "none"
}

/** One story (ours or someone else's). Times in epoch MILLIS. */
data class Story(
    val id: String,
    val authorId: String,
    val isMine: Boolean,
    val createdAt: Long,
    val expiresAt: Long,
    /** The R2 object key + media key + nonce + ciphertext hash. The key stops on-device. */
    val media: ChatEngine.MediaRef,
    val caption: String,
    val durationMs: Long?,
    val width: Int?,
    val height: Int?,
    val allowsReplies: Boolean,
    /** Epoch MILLIS, null = unseen. Recorded on this device only, never transmitted. */
    val viewedAt: Long?,
    /** Absolute path to the decrypted plaintext file, once downloaded. */
    val localPath: String?,
    val downloadState: StoryDownloadState,
    val uploadState: StoryUploadState,
) {
    val isImage: Boolean get() = media.mime.startsWith("image/")
    val isVideo: Boolean get() = media.mime.startsWith("video/")
    fun isExpired(nowMs: Long = System.currentTimeMillis()): Boolean = expiresAt <= nowMs
}

/**
 * A *context* — one author's set of unexpired stories, the unit the tray and the outer viewer
 * pager work in (one row / one page per author, never one per story).
 */
data class StoryContext(
    val authorId: String,
    val authorName: String,
    val photoUrl: String?,
    val stories: List<Story>,
) {
    val isMine: Boolean get() = stories.firstOrNull()?.isMine == true
    val hasUnviewed: Boolean get() = stories.any { it.viewedAt == null }
    val newest: Story? get() = stories.maxByOrNull { it.createdAt }
    /** The next story to auto-advance to: first unviewed, else the first. */
    val startIndex: Int get() = stories.indexOfFirst { it.viewedAt == null }.let { if (it < 0) 0 else it }
}

/** One viewer of one of OUR stories (author-side, from decrypted receipts). */
data class StoryViewer(val userId: String, val name: String, val viewedAtMs: Long)

// MARK: - Wire envelopes (the AUTHENTICATED PLAINTEXT of a ratchet message)
//
// There is NO AAD parameter on any AEAD in e2e-core, so story_id / author / expiry cannot be
// bound at the AEAD layer — they go INSIDE the plaintext, where GCM authenticates them. Receivers
// MUST validate them (see StoryEngine.validate). Field names match the iOS Codable + the spec.

@Serializable
data class StoryEnvelope(
    val v: Int? = 1,
    val t: String? = "story",
    val story_id: String,
    val author_id: String,
    val created_at: Long,       // epoch MILLIS
    val expires_at: Long,       // epoch MILLIS (author's claim)
    val media: ChatEngine.MediaRef,   // the SAME MediaRef shape used for chat media — the real mime lives here
    // NULLABLE, not just defaulted. Kotlin defaults cover an ABSENT key, but iOS's
    // JSONEncoder emits `"caption": null` explicitly for a nil Optional, and kotlinx
    // throws on an explicit null for a non-nullable property. That threw on every
    // iOS→Android moment and the story was dropped for good (the feed is deliver-once).
    // The mirror of the Android→iOS bug already fixed in iOS's Story.swift.
    val caption: String? = "",
    val durationMs: Long? = null,
    val width: Int? = null,
    val height: Int? = null,
    val allowsReplies: Boolean? = true,
)

/** A reply to a story — sent as an ORDINARY 1:1 message with content_type "story_reply". */
@Serializable
data class StoryReplyEnvelope(
    val v: Int = 1,
    val t: String = "story_reply",
    val storyId: String,
    val storyAuthorId: String,
    val storyCreatedAt: Long,
    val text: String,
    val reaction: String? = null,
)

/** A view receipt — encrypted once per AUTHOR device. The viewer id is IN here, not in a column. */
@Serializable
data class StoryViewReceipt(
    val v: Int = 1,
    val t: String = "story_view",
    val story_id: String,
    val viewer_id: String,
    val viewed_at: Long,        // epoch MILLIS
)
